### PR TITLE
Ja/move api clients to docstrings

### PR DESCRIPTION
**Pull request by monkey-jeff**
Work item: W-7769115

This does a few things in an attempt to simplify the code before splitting up

1. APICLIENTS.md and templates to create them have been removed
2. The `listOperations` has been removed as that wasn't needed for anything except a demo of what could be done.
3. *MAJOR CHANGE* Rendering of API models refactored to allow for the passing of the API metadata from exchange.
4. A lot of unneeded code and their tests have been removed.
5. Adds package-lock.json back in (If you ignore this there are only 91 lines added, and over 300 removed in this PR)
6. Class docstrings now have examples and links back to their corresponding documentation at https://developer.commercecloud.com/ 

Example output of new documentation generated - 
Screen Shot 2020-07-04 at 3 05 12 PM

### CODE REVIEW

File: build-config.ts [hunk 1]
```diff
@@ -9,6 +9,8 @@ export default {
   inputDir: process.env.COMMERCE_SDK_INPUT_DIR || `${__dirname}/apis`,
   updateApiDir: process.env.COMMERCE_SDK_UPDATE_API_DIR || `${__dirname}/updateApiTmpDir`,
   renderDir: process.env.COMMERCE_SDK_RENDER_DIR || `${__dirname}/packages/generator/renderedTemplates`,
+  
+  // TODO: This this should become 'group by' and provide a method to pickout the metadata you want to group by
```
Comment (dbarbato-sf) [inline build-config.ts:13]: Please add a story for the todo.

File: packages/generator/src/renderer.ts [hunk 2]
```diff
@@ -155,15 +135,19 @@ export function loadApiConfig(
 export async function processApiFamily(
   apiFamily: RestApi[],
   inputDir: string
-): Promise<model.document.Document[]> {
+  // TODO: Once logic is moved to raml-toolkit this should just be an optional field on RestApi
```
Comment (dbarbato-sf) [inline packages/generator/src/renderer.ts:138]: Please add this to that story and add a story number here for tracking

File: packages/generator/src/renderer.ts [hunk 3]
```diff
@@ -224,13 +189,15 @@ export function objectFromEntries<T>(
  */
 function createClient(
   webApiModel: model.document.BaseUnitWithDeclaresModel,
-  apiName: string
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  apiMetadata: { [key: string]: any }
 ): string {
+  // console.log(apiMetadata);
```
Comment (dbarbato-sf) [inline packages/generator/src/renderer.ts:195]: Delete line

File: build-config.ts (hunk 1 above)
Comment (monkey-jeff) [reply build-config.ts:13]: There already is one.  I am pretty sure @bredmond-sf wrote a story for this in his migration plan. I know I have discussed it with him.  Was adding this as a kind of marker to start talking about what that would look like.

File: packages/generator/src/renderer.ts (hunk 3 above)
Comment (monkey-jeff) [reply packages/generator/src/renderer.ts:195]: Oops, missed one!

File: packages/generator/src/renderer.ts (hunk 2 above)
Comment (monkey-jeff) [reply packages/generator/src/renderer.ts:138]: Actually no,  going to remove this line :-p This was actually for a new parameter for renderer that I ended up deleting as it wasn't needed yet for my story.  So Ill just remove this comment.

File: build-config.ts (hunk 1 above)
Comment (monkey-jeff) [reply build-config.ts:13]: This also kind of serves as instructions for what to do with that line when you work on it.

File: packages/generator/src/renderer.ts [hunk 4]
```diff
@@ -52,18 +52,23 @@ interface IBuildConfig {
...
-  family: string;
-  config: RestApi;
-}
+export type DocumentWithMetadataT = {
+  document: model.document.Document;
+  metadata: {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    [key: string]: any;
+  };
+};
```
Comment (wjhsf) [inline packages/generator/src/renderer.ts:66]:
```suggestion
type DocumentMetadata = RestApi & {
  apiSpec: string;
  specName: string;
};
export type DocumentWithMetadataT = {
  document: model.document.Document;
  metadata: DocumentMetadata;
};
```

File: packages/generator/test/renderers/helpers.test.ts [hunk 5]
```diff
@@ -98,7 +52,13 @@ describe("Renderer Helpers", () => {
       expect(models)
         .to.be.an("array")
         .with.lengthOf(1);
-      expect(models[0]).to.be.an.instanceOf(model.document.Document);
```
Comment (wjhsf) [inline packages/generator/test/renderers/helpers.test.ts:101]: This test should still be included, just change the expect to `models[0].document`.

File: packages/generator/test/renderers/helpers.test.ts [hunk 6]
```diff
@@ -127,36 +87,7 @@ describe("Renderer Helpers", () => {
...
-      const nested = { property: "value" };
-      const method = (): null => null;
-      const result = renderer.objectFromEntries([
-        ["nested", nested],
-        ["method", method]
-      ] as Entries);
-      expect(result).to.deep.equal({ nested, method });
-      expect(result.nested).to.equal(nested);
-      expect(result.method).to.equal(method);
+      expect(result[0][1][0]).to.have.keys(["document", "metadata"]);
```
Comment (wjhsf) [inline packages/generator/test/renderers/helpers.test.ts:90]: The deleted test should remain, and an additional test should be added to test the expected structure of `metadata`.

File: packages/generator/test/renderers/renderers.test.ts [hunk 7]
```diff
@@ -88,51 +88,4 @@ describe("Renderers", () => {
       );
     }).timeout(10000);
   });
-
```
Comment (wjhsf) [inline packages/generator/test/renderers/renderers.test.ts:91]: `js-yaml` is no longer needed. It should be removed from this file and removed as a dependency.

File: packages/generator/src/renderer.ts (hunk 2 above)
Comment (bredmond-sf) [reply packages/generator/src/renderer.ts:138]: This function will go away in the story to use the raml-toolkit refactor anyway. ApiGroup.read() pretty much handles all of this. This looks good right now.

File: packages/generator/src/renderer.ts (hunk 4 above)
Comment (monkey-jeff) [reply packages/generator/src/renderer.ts:66]: I intentionally left metadata to be whatever you wanted.  Did NOT want to specify specific constraints here.

File: packages/generator/src/renderer.ts [hunk 8]
```diff
@@ -52,18 +52,23 @@ interface IBuildConfig {
...
- * Information used to generate APICLIENTS.md.
+ * Api Document with metadata
+ * TODO: This should be moved into raml-toolkit
  */
-export interface IApiClientsInfo {
-  model: model.domain.WebApi;
-  family: string;
-  config: RestApi;
-}
+export type DocumentWithMetadataT = {
```
[the remark this answers is not in the document]
Comment (monkey-jeff) [reply packages/generator/src/renderer.ts:60]: When having `[key: string]: any;` it must be a type.  I also think until we have a story talking about the structure of metadata this makes more sense.

File: packages/generator/src/renderer.ts (hunk 2 above)
Comment (monkey-jeff) [reply packages/generator/src/renderer.ts:138]: That's what I figured. Which is why the comment specifically says "move the logic" I don't care about the function at all.

File: packages/generator/test/renderers/helpers.test.ts (hunk 6 above)
Comment (monkey-jeff) [reply packages/generator/test/renderers/helpers.test.ts:90]: You mean the deleted assertion right? These tests should be deleted as the methods they were testing were removed.

File: packages/generator/src/renderer.ts [hunk 9]
```diff
@@ -224,13 +189,15 @@ export function objectFromEntries<T>(
  */
 function createClient(
   webApiModel: model.document.BaseUnitWithDeclaresModel,
-  apiName: string
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  apiMetadata: { [key: string]: any }
```
[the remark this answers is not in the document]
Comment (monkey-jeff) [reply packages/generator/src/renderer.ts:193]: I'd rather not prescribe what metadata can contain..yet.

File: packages/generator/src/renderer.ts [hunk 10]
```diff
@@ -345,25 +271,25 @@ export function createOperationList(allApis: {
...
-function renderApi(
-  apiModel: model.document.Document,
-  renderDir: string
-): string {
-  const apiName: string = getApiName(apiModel);
-  const apiPath: string = path.join(renderDir, apiName);
-  fs.ensureDirSync(apiPath);
+function renderApi(apiModel: DocumentWithMetadataT, renderDir: string): string {
+  apiModel.metadata.specName = getApiName(apiModel.document);
+  apiModel.metadata.apiPath = path.join(renderDir, apiModel.metadata.specName);
```
[the remark this answers is not in the document]
Comment (monkey-jeff) [reply packages/generator/src/renderer.ts:274]: It does...but decided that the refactor Brian is working will supersede this code anyways.  Also one of the reasons I wanted to keep metadata fluid.  At any point, we have the data we want to add I wanted to add it.

File: packages/generator/src/renderer.ts (hunk 4 above)
Comment (wjhsf) [reply packages/generator/src/renderer.ts:66]:
```suggestion
type DocumentMetadata = RestApi & {
  apiSpec: string;
  specName: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [key: string]: any;
};
export type DocumentWithMetadataT = {
  document: model.document.Document;
  metadata: DocumentMetadata;
};
```

With this, known properties are typed while still allowing extensibility.

**Comment by monkey-jeff** [comment]
Finally fixed the snyk issues!!!

File: packages/generator/src/renderer.ts (hunk 4 above)
Comment (monkey-jeff) [reply packages/generator/src/renderer.ts:66]: Not terrible worried about this here.  This is definitely a nice enhancement once we move stuff over.  But would rather keep this as simple as possible at this point.  apiSpec and apiName aren't freeform metadata in Brian's new PR, they are known aspects of the class.  Basically would rather keep what is here simpler to make the transition easier.

File: packages/generator/src/renderer.ts [hunk 11]
```diff
@@ -52,18 +52,23 @@ interface IBuildConfig {
...
-export interface IApiClientsInfo {
-  model: model.domain.WebApi;
-  family: string;
-  config: RestApi;
-}
+export type DocumentWithMetadataT = {
+  document: model.document.Document;
+  metadata: {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    [key: string]: any;
```
Comment (unandyala) [inline packages/generator/src/renderer.ts:64]: consider using unknown - type-safe counterpart of any

Comment (monkey-jeff) [reply packages/generator/src/renderer.ts:64]: Unkown didn't work due to something else down stream.  And really just trying to remove logic for the migration

**Pull request merged**